### ISSUE TITLE
Imports may get removed if using Prettier range-start/range-end options

### ISSUE THREAD

**Issue by alexander-doroshko** [simonhaenisch/prettier-plugin-organize-imports#110]
**Steps to reproduce**
- Create a trivial project with a few files:

`package.json:`
```
{
  "devDependencies": {
    "prettier": "3.0.3",
    "prettier-plugin-organize-imports": "3.2.3"
  }
}
```
`.prettierrc:`
```
{
  "plugins": [
    "prettier-plugin-organize-imports"
  ]
}
```
`foo.ts:`
```
export const foo = 0;
```
`bar.ts:`
```
import {foo} from "./foo";

let bar = foo;
```

- Run `prettier --range-end=27 bar.ts`. (In fact, any range end between 1 and 30 will demonstrate the bug)

**Expected result**
- No changes in file

**Actual result**
- Import statement disappears:
```


let bar = foo;
```

The problem is crucial for WebStrom and other JetBrains IDEs because Prettier integration in JetBrains IDEs ask Prettier to format only the imports block pretty often.

**Comment by simonhaenisch** [comment]
Very interesting, thanks for sharing. Not sure what can be done about this implementation-wise though, since by passing only a range to Prettier you are taking away the whole context of the file which obviously is needed for the TypeScript API to figure out which imports are or aren't unused.

In this use-case you'll just have to turn on the [skip destructive code actions](https://github.com/simonhaenisch/prettier-plugin-organize-imports#skip-destructive-code-actions) option.

Or request an option in those IDE's Prettier integrations to not do a partial format.

**Comment by MattLishmanYara** [comment]
@simonhaenisch Would it be possible to just have this plugin do nothing when a range is passed?
While the problem shows itself in certain IDEs, I would say this is a bug in the plugin to be destructive when using prettier to format a range of a file.

**Comment by alexander-doroshko** [comment]
@simonhaenisch Thanks for the answer.

Is it really not possible for the `prettier-plugin-organize-imports` plugin to get the full context of a file when formatting a range?

If so, I'd also be happy if the plugin does nothing.

**Comment by simonhaenisch** [comment]
> Is it really not possible for the `prettier-plugin-organize-imports` plugin to get the full context of a file when formatting a range?

Yeah sure could probably manually read the file but then you might as well just tell your IDE to format the whole document not just the imports? 🤷🏻‍♂️ anyway seems like a bad thing to implement when Prettier already gives you the code it's supposed to format (seems fragile, creates performance overhead and maintenance cost).

> If so, I'd also be happy if the plugin does nothing.

Then either enable the option I mentioned or remove the plugin?

**Comment by simonhaenisch** [comment]
Or do you mean the plugin should only do nothing if a range is provided? Could make sense

**Comment by MattLishmanYara** [comment]
> Or do you mean the plugin should only do nothing if a range is provided? Could make sense

I agree with this too. I think the plugin should just do nothing if a range is passed.

**Comment by simonhaenisch** [comment]
Alright, makes sense. Or at least automatically force-enabling `noDestructiveCodeActions` if only a range is formatted, since in a partial format it's kinda guaranteed that it would misbehave.

Anyway, problem: it seems that `rangeStart` and `rangeEnd` are always set, even when formatting the whole file 😕 

@alexander-doroshko just wondering, why is the order not just

 - optimize imports using IDE's own engine
 - format the whole file using Prettier

Makes a lot more sense to me and prevents the issue in the first place, plus you would be faster because don't double-format a part of the document.

**Comment by alexander-doroshko** [comment]
> just wondering, why is the order not just...

The current implementation was just the easiest and the most logical way to add Prettier integration to the IDE. The IDE has its own smartness - its own code formatting and import optimization tools. So, it has its order of actions. The easiest way to add Prettier integration was to catch the moments when the IDE wanted to format the whole file or a range and to call Prettier for this file or this range. This way we've got such an order of actions. Any other solution would require significant changes in the IDE engine.

**Comment by simonhaenisch** [comment]
> Anyway, problem: it seems that `rangeStart` and `rangeEnd` are always set, even when formatting the whole file 😕

Do you have an idea how to figure out if the plugin received a range or the full file, given that it seems to always set `rangeEnd`?

**Comment by alexander-doroshko** [comment]
> Do you have an idea how to figure out if the plugin received a range or the full file

Sorry, no idea, I've never been writing a prettier plugin

**Comment by MattLishmanYara** [comment]
> > Anyway, problem: it seems that `rangeStart` and `rangeEnd` are always set, even when formatting the whole file 😕
> 
> Do you have an idea how to figure out if the plugin received a range or the full file, given that it seems to always set `rangeEnd`?

@simonhaenisch 

As the `code` is just a string and the `rangeEnd` is just the length of the string, isn't it enough to just check if the `rangeEnd` !== `code.length`?

I just did some very quick testing of this and it seems to solve the issue

```
const organizeImports = (code, options) => {
	if (
		code.includes('// organize-imports-ignore') ||
		code.includes('// tslint:disable:ordered-imports') ||
		code.length !== options.rangeEnd ||
		options.rangeStart !== 0
	) {
		return code;
	}
```

**Comment by MattLishmanYara** [comment]
I can try to add some tests and submit a PR tonight

**Comment by MattLish** [comment]
@simonhaenisch This should solve the issue with `rangeEnd` when using the CLI https://github.com/simonhaenisch/prettier-plugin-organize-imports/pull/112

@alexander-doroshko This solves the `rangeEnd` issue via the CLI but doesn't seem to solve the IDE issue.
When testing the original example from this issue in WebStorm, I don't see the imports disappear when running a reformat, with or without my fix.

When I use some slightly longer code, I see the issue in WebStorm with and without my fix.
Is WebStorm caching the node_modules? 
Or doing something else?
```
import { compare } from 'fast-json-patch';

const x = 100;

function y() {
  compare({}, {});
}
```

**Comment by alexander-doroshko** [comment]
Close/open the project in the IDE to make sure there are no caching issues

**Comment by MattLish** [comment]
> Close/open the project in the IDE to make sure there are no caching issues

@alexander-doroshko Ah, that was the issue 😅 
After closing and reopening, I can confirm this PR fixes the issue experienced in the IDE 👍 

@simonhaenisch Looks like the PR is sufficient, I would appreciate it if you could review it when you have time please 😄 .

**Comment by simonhaenisch** [comment]
Hey thanks for the PR, will try to get to it soon!

Kinda weird though that `code.length` would be larger than `rangeEnd`, I was assuming that it would actually only pass the code of the affected range to the plugin 🤔 because if `code` contains the whole file, then doing a partial format shouldn't actually remove seemingly unused imports?

**Comment by MattLishmanYara** [comment]
That's an interesting point actually.
From what I could tell, the `organizeImports` method seemed to be called twice.
Once with the full code and then again just with the range.
I can't see this in prettier docs anywhere.

I _guess_ this could be a way to gather the context of the whole file and then just format a range.
However, I would suspect most people using this plugin aren't using `rangeEnd` and it's just suitable to do nothing destructive if it is used.

**Comment by simonhaenisch** [comment]
I'm gonna try to find some time to double-check your PR because if `code` contains the full file, then the plugin should do the right thing (not remove imports) even when a range is set (since the plugin just passes whatever `code` to the TS API).

**Comment by simonhaenisch** [comment]
@alexander-doroshko fixed in `3.2.4`, it now completely skips when formatting a range, which is based on the following conditions:

- either `options.originalText` is truthy
- or `options.rangeStart !== 0`
- or `options.rangeEnd !== Infinity` and `options.rangeEnd !== code.length`

**Comment by alexander-doroshko** [comment]
Cool! Thanks a lot @simonhaenisch, @MattLish, and @MattLishmanYara :)
